src/rdma.c

```diff
@@ -0,0 +1,985 @@
+    conn_param.responder_resources = 1;
+    conn_param.initiator_depth = 1;
+    conn_param.retry_count = 7;
+    conn_param.rnr_retry_count = 7;
+    if (rdma_connect(cm_id, &conn_param)) {
+        valkeySetError(c, VALKEY_ERR_OTHER, "RDMA: connect failed");
+        goto destroy_iobuf;
+    }
+
+    return VALKEY_OK;
+
+destroy_iobuf:
+    rdmaDestroyIoBuf(ctx);
+free_qp:
+    ibv_destroy_qp(cm_id->qp);
+error:
+    if (cq)
+        ibv_destroy_cq(cq);
+    if (pd)
+        ibv_dealloc_pd(pd);
+    if (comp_channel)
+        ibv_destroy_comp_channel(comp_channel);
+
+    return VALKEY_ERR;
+}
+
+static int ValkeyRdmaEstablished(valkeyContext *c, struct rdma_cm_id *cm_id) {
+    RdmaContext *ctx = (RdmaContext *)c->privctx;
+
+    /* it's time to tell redis we have already connected */
+    c->flags |= VALKEY_CONNECTED;
+    c->funcs = &valkeyContextRdmaFuncs;
+    c->fd = ctx->comp_channel->fd;
+
+    return connRdmaRegisterRx(ctx, cm_id);
+}
+
+static int ValkeyRdmaCM(valkeyContext *c, int timeout) {
+    RdmaContext *ctx = (RdmaContext *)c->privctx;
+    struct rdma_cm_event *event;
+    char errorstr[128];
+    int ret = VALKEY_ERR;
+
+    while (rdma_get_cm_event(ctx->cm_channel, &event) == 0) {
+        switch (event->event) {
+        case RDMA_CM_EVENT_ADDR_RESOLVED:
+            if (timeout < 0 || timeout > 100)
+                timeout = 100; /* at most 100ms to resolve route */
+            ret = rdma_resolve_route(event->id, timeout);
+            if (ret) {
+                valkeySetError(c, VALKEY_ERR_OTHER, "RDMA: route resolve failed on");
+            }
+            break;
+        case RDMA_CM_EVENT_ROUTE_RESOLVED:
+            ret = ValkeyRdmaConnect(c, event->id);
+            break;
+        case RDMA_CM_EVENT_ESTABLISHED:
+            ret = ValkeyRdmaEstablished(c, event->id);
+            break;
+        case RDMA_CM_EVENT_TIMEWAIT_EXIT:
+            ret = VALKEY_ERR;
+            valkeySetError(c, VALKEY_ERR_TIMEOUT, "RDMA: connect timeout");
+            break;
+        case RDMA_CM_EVENT_ADDR_ERROR:
+        case RDMA_CM_EVENT_ROUTE_ERROR:
+        case RDMA_CM_EVENT_CONNECT_ERROR:
+        case RDMA_CM_EVENT_UNREACHABLE:
+        case RDMA_CM_EVENT_REJECTED:
+        case RDMA_CM_EVENT_DISCONNECTED:
+            c->flags &= ~VALKEY_CONNECTED;
+            break;
+        case RDMA_CM_EVENT_ADDR_CHANGE:
+        default:
+            snprintf(errorstr, sizeof(errorstr), "RDMA: connect failed - %s", rdma_event_str(event->event));
+            valkeySetError(c, VALKEY_ERR_OTHER, errorstr);
+            ret = VALKEY_ERR;
+            break;
+        }
+
+        rdma_ack_cm_event(event);
+    }
+
+    return ret;
+}
+
+static int ValkeyRdmaWaitConn(valkeyContext *c, long timeout) {
+    int timed;
+    struct pollfd pfd;
+    long now = vk_msec_now();
+    long start = now;
+    RdmaContext *ctx = (RdmaContext *)c->privctx;
+
+    while (now - start < timeout) {
+        timed = (int)(timeout - (now - start));
```

Comment:
The casting to a smaller integer type makes me somewhat nervous :smile:

What about reversing the formulation from a `while (now - start < timeout) { ... }`

By calculating the end time up front (rough sketch)

```c
    assert (timeout >= 0);

    long now = vk_msec_now(), end = now + timeout;
    do {
        /* ... */
        if (poll(&pfd, 1, end - now) < 0) { }
        /* ... */
        if (ValkeyRdmaCM(c, end - now) {  }
        now = vk_msec_now();
    } while(now < end);
```